### DISCUSSION TITLE
CORS Requests + Responses

### DISCUSSION THREAD

**Discussion by gxxcastillo** [mswjs/msw#2040] [Q&A] [answered]
When running my integration tests, which use MSW for mocking api requests, I'm noticing "OPTIONS" pre-flight requests are being made.

What's the recommended approach for dealing with these CORS requests?

I've gone about adding the following handler: 
```
  http.all('*', ({ request }) => {
    if (request.method === 'OPTIONS') {
      return new Response(null, {
        status: 204,
        headers: {
          'Access-Control-Allow-Origin': '*',
          'Access-Control-Allow-Methods': 'GET,PUT,POST,DELETE',
          'Access-Control-Allow-Headers': '*',
        },
      });
    }
  ...
```

However, I'm still getting Network Errors.  Is there something I need to do for the MSW server to handle CORS requests?

**Comment by kettanaito** [comment, marked as answer]
Hi, @gxxcastillo. 

What you are doing looks correct to me. You can use `http.options()` handler to add the CORS response to all requests or a subset of requests.

```js
http.options('*', () => {
  return new Response(null, {
    status: 200,
    headers: {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'GET,PUT,POST,DELETE',
      'Access-Control-Allow-Headers': '*',
    },
  })
})
```

> However, I'm still getting Network Errors.

Can you please share those errors? Where are they coming from? What do they state?

**Comment by gxxcastillo** [comment]
Hi, just following up on this thread.  I haven't seen the CORS errors in a while now and I haven't had to use the http.options() handler.  The errors seem to have gone away on their own.

**Reply by kettanaito** [reply]
CORS is usually triggered by your request client (or the browser). Perhaps something changed, and that resulted in the issue auto-fixing itself. 

If anyone encounters CORS errors, observe the traffic you have, and make sure to provide the respective request handlers for CORS requests.

**Comment by piotr-cz** [comment]
I'm interested in opposite behavior:

I'd like to test CORS preflight requests with MSW, but somehow Chrome 137 doesn't use them when MSW is enabled, even for cros-origin requests.

**Reply by kettanaito** [reply]
Can you please provide the code for this? Ideally, with the request handlers how you expect them to behave. Keep in mind that you're simulating server behaviors with MSW, not the client ones.

**Reply by piotr-cz** [reply]
I'd like to use similar code in handlers as you provided above: https://github.com/mswjs/msw/discussions/2040#discussioncomment-8925667 to handle preflight requests.

Probably there should be a configuration option in the `worker.start` method to trigger CORS (if it's possible)
